jscomp/others/js_json.ml

```diff
@@ -129,3 +129,61 @@ external stringify: t -> string = "stringify"
   [@@bs.val] [@@bs.scope "JSON"]
 external stringifyWithSpace: t -> (_ [@bs.as {json|null|json}]) -> int -> string = "stringify"
   [@@bs.val] [@@bs.scope "JSON"]
+
+
+(* in memory modification does not work until your root is
+   actually None, so we need wrap it as `[v]` and
+   return the first element instead *)
+
+let patch : _ -> _ = [%raw{|function (json) {
+  var x = [json];
+  var q = [{ kind: 0, i: 0, parent: x }];
+  while (q.length !== 0) {
+    // begin pop the stack
+    var cur = q[q.length - 1];
+    if (cur.kind === 0) {
+      cur.val = cur.parent[cur.i]; // patch the undefined value for array
+      if (++cur.i === cur.parent.length) {
+        q.pop();
+      }
+    } else {
+      q.pop();
+    }
+    // finish
+    var task = cur.val;
+    if (typeof task === "object") {
+      if (Array.isArray(task) && task.length !== 0) {
+        q.push({ kind: 0, i: 0, parent: task, val: undefined });
+      } else {
+        for (var k in task) {
+          if (k === "RE_PRIVATE_NONE") {
+            if (cur.kind === 0) {
+              cur.parent[cur.i - 1] = undefined;
+            } else {
+              cur.parent[cur.i] = undefined;
+            }
+            continue;
+          }
+          q.push({ kind: 1, i: k, parent: task, val: task[k] });
+        }
+      }
+    }
+  }
+  return x[0];
+}
+|}]
+
+
+let serializeExn (type t) (x : t) : string option = [%raw{| function(obj){
+  return JSON.stringify(obj,function(_,value){
+      if(value===undefined){
+          return {RE_PRIVATE_NONE : true}
+      }
+    return value
+  })
+  }
+|}] x 
+
+let deserializeExn (s: string) : 'a = 
+  patch (parseExn s)
```

Comment:
Is there a reason why you don't use the second `JSON.parse` argument like the following?

```js
JSON.parse(s, function(_key, value) {
  if(value && value.RE_PRIVATE_NONE === true) {
    return undefined
  };
  return value
});
```

Reply:
Unfortunately this does not work, you can try it

Reply:
<img width="732" alt="Screenshot 2020-06-06 at 14 00 05" src="https://user-images.githubusercontent.com/1688645/83943636-07a4e400-a7fe-11ea-81b0-dd7fcc61bb65.png">

Tried on Chrome and Safari and it does seem to work? The only difference I see is that it makes the arrays sparse, but that doesn't seem to be an issue given how BS's runtime interacts with them?

Reply:
Oh yeah, that'd break `Js.Array.*`. Nevermind 😄

Reply:
Something like this would work, I'm not sure if that'd optimise better in browser VMs

```js
let pairs = [];
let parsed = JSON.parse(s, function(key, value) {
  if(value && value.RE_PRIVATE_NONE === true) {
    pairs.push([this, key]);
    return undefined
  };
  return value
});
pairs.forEach(([obj, key]) => obj[key] = undefined)
parsed;
```

Reply:
it does not work for me, do you have a working example? is `this` semantics documented somewhere?

Even if it works, it seems the IC is still changed

Reply:
```js
let p = s => {var parents = [];
var obj = JSON.parse(s, function(key, value) {
  if(value && value.RE_PRIVATE_NONE === true) {
    parents.push({par:this, key});
  };
  return value
}); for(let {par,key} of parents){par[key]=undefined}; return obj }
```
This looks good to me

Reply:
Documented here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#Using_the_reviver_parameter

I made a rough benchmark in Chrome and Safari and it seems that your original solution is consistently faster, so the only benefit left would be runtime size. About IC it seems like returning `undefined` from the reviver function will necessarily call some `delete`, but I'm wondering if using `null` as a placeholder until replacing "by hand" with `undefined` would help.

Reply:
It maybe that JSON.parse without reviver is specialised. 
About the bundle size, my implementation is quite small, if it is indeed faster, we can keep it.
It is good to know an alternative solution!

Reply:
I am thinking we can do the same thing with stringify without using callback, it may be faster